plugins/http_plugin/beast_http_session.hpp

```diff
@@ -0,0 +1,511 @@
+                , stream_(std::move(socket))
+            {
+            }
+
+            // Called by the base class
+            beast::tcp_stream&
+            stream()
+            {
+                return stream_;
+            }
+
+            // Start the asynchronous operation
+            void
+            run()
+            {
+                // Set the timeout.
+                beast::get_lowest_layer(stream_).expires_after(std::chrono::seconds(30));
+
+                // We need to be executing within a strand to perform async operations
+                // on the I/O objects in this session. Although not strictly necessary
+                // for single-threaded contexts, this example code is written to be
+                // thread-safe by default.
+                asio::dispatch(stream_.get_executor(),
+                            beast::bind_front_handler(
+                                &plain_session::do_read,
+                                shared_from_this()));
+            }
+
+            void
+            do_eof()
+            {
+                // Send a TCP shutdown
+                beast::error_code ec;
+                stream_.socket().shutdown(tcp::socket::shutdown_send, ec);
+
+                // At this point the connection is closed gracefully
+            }
+
+            bool is_secure() { return false; }
+
+    }; // end class plain_session
+
+    // Handles an SSL HTTP connection
+    class ssl_session
+        : public beast_http_session<ssl_session>
+        , public std::enable_shared_from_this<ssl_session>
+    {
+        // std::unique_ptr<beast::ssl_stream<beast::tcp_stream> > stream_;
+        beast::ssl_stream<beast::tcp_stream> stream_;
+
+        public:
+            // Create the session
+            ssl_session(
+                tcp::socket&& socket,
+                std::shared_ptr<ssl::context> ctx,
+                std::shared_ptr<http_plugin_state> plugin_state,
+                asio::io_context* ioc)
+                : beast_http_session<ssl_session>(plugin_state, ioc),
+                    stream_(std::move(socket), *ctx)
+            { }
+
+            // Called by the base class
+            beast::ssl_stream<beast::tcp_stream>&
+            stream()
+            {
+                return stream_;
+            }
+
+            // Start the asynchronous operation
+            void
+            run()
+            {
+                
+                // Set the timeout.
+                // beast::get_lowest_layer(stream_).expires_after(std::chrono::seconds(30));
+
+                auto self = shared_from_this();
+
+            
+                // We need to be executing within a strand to perform async operations
+                // on the I/O objects in this session.
+                asio::dispatch(stream_.get_executor(), [self]() {
+                    // Set the timeout.
+                    
+                    beast::get_lowest_layer(self->derived().stream_).expires_after(
```

Comment:
`expires_after()` has been notoriously problematic for us in other areas, @heifner can probably elaborate more.

Reply:
@heifner Could you elaborate on problem with `expires_after()` ?

Reply:
https://github.com/EOSIO/eos/pull/10252